// elysia-session/dist/stores/bun/sqlite.js
export class BunSQLiteStore {
    constructor(db, tableName) {
        this.db = db;
        this.tableName = tableName;
        this.db.query(`CREATE TABLE IF NOT EXISTS ${this.tableName} (id TEXT PRIMARY KEY, data TEXT)`).run();
    }
    getSession(id, ctx) {
        if (!id)
            return null;
        const query = this.db.query(`SELECT data FROM ${this.tableName} WHERE id = $id`);
        const result = query.get({ $id: id });
        if (!result)
            return null;
        // @ts-expect-error - data property is not defined in type
        return JSON.parse(result.data);
    }
    createSession(data, id, ctx) {
        const query = this.db.query(`INSERT INTO ${this.tableName} (id, data) VALUES ($id, $data)`);
        query.run({ $id: id, $data: JSON.stringify(data) });
    }
    deleteSession(id, ctx) {
        if (!id)
            return;
        const query = this.db.query(`DELETE FROM ${this.tableName} WHERE id = $id`);
        query.run({ $id: id });
    }
    persistSession(data, id, ctx) {
        if (!id)
            return;
        const query = this.db.query(`UPDATE ${this.tableName} SET data = $data WHERE id = $id`);
        query.run({ $id: id, $data: JSON.stringify(data) });
    }
}
